samza-core/src/main/scala/org/apache/samza/storage/ContainerStorageManagerRestoreUtil.java

```diff
@@ -156,53 +158,90 @@ private static CompletableFuture<Map<TaskName, Checkpoint>> restoreAllTaskInstan
+      SamzaContainerMetrics samzaContainerMetrics, TaskName taskInstanceName) {
+    return restoreFuture.handle((aVoid, ex) -> {
+      if (ex != null) {
+        if (isUnwrappedExceptionDeletedException(ex)) {
+          // if the exception is of type DeletedException, retry restore (with getDeleted flag set to true).
+          return true;
+        } else {
           updateRestoreTime(startTime, samzaContainerMetrics, taskInstanceName);
+          // log and rethrow exception to communicate restore failure
+          String msg = String.format("Error restoring state for task: %s", taskInstanceName.getTaskName());
+          LOG.error(msg, ex);
+          throw new SamzaException(msg, ex); // wrap in unchecked exception to throw from lambda
+        }
+      }
+      return false;
+    });
+  }
 
-          if (ex != null) {
-            if (isUnwrappedExceptionDeletedException(ex)) {
-              LOG.warn("Received DeletedException during restore for task {}. Attempting to get blobs with getDeleted flag",
-                  taskInstanceName.getTaskName());
-
-              // Try to restore with getDeleted flag
-              CompletableFuture<Checkpoint> future =
-                  restoreDeletedSnapshot(taskInstanceName, taskCheckpoints,
-                      checkpointManager, taskRestoreManager, config, taskInstanceMetrics, executor,
-                      taskBackendFactoryToStoreNames.get(taskInstanceName).get(factoryName), loggedStoreDir,
-                      jobContext, containerModel);
-              try {
-                newTaskCheckpoints.put(taskInstanceName, future);
-              } catch (Exception e) {
-                String msg = String.format("DeletedException during restore task: %s after retrying to get deleted blobs.", taskName);
-                throw new SamzaException(msg, e);
-              } finally {
-                updateRestoreTime(startTime, samzaContainerMetrics, taskInstanceName);
-              }
-            } else {
-              // log and rethrow exception to communicate restore failure
-              String msg = String.format("Error restoring state for task: %s", taskName);
-              LOG.error(msg, ex);
-              throw new SamzaException(msg, ex); // wrap in unchecked exception to throw from lambda
-            }
-          }
+  private static CompletableFuture<Checkpoint> retryFailedRestore(CompletableFuture<Boolean> shouldRetryRestoreFuture,
+      TaskName taskName, Map<TaskName, Checkpoint> taskCheckpoints, CheckpointManager checkpointManager,
+      TaskRestoreManager taskRestoreManager, Config config, Map<TaskName, TaskInstanceMetrics> taskInstanceMetrics,
+      ExecutorService executor, File loggedStoreBaseDirectory, JobContext jobContext, ContainerModel containerModel,
+      String factoryName, Map<TaskName, Map<String, Set<String>>> taskBackendFactoryToStoreNames) {
+    return shouldRetryRestoreFuture.thenCompose(shouldRetryRestore -> {
```

Comment:
Blocker: Extract the contents of the lambdas to methods, and keep the future chaining in the caller.
Makes it easier to understand the execution plan.